File: web_estatico/js/motor.js
// Puerto a JavaScript del motor de recomendacion (src/motor/motor.py).
// Misma especificacion, mismas reglas, sin ninguna desviacion.
// Se valida contra el motor Python real mediante tests_js/test_paridad.js.
//
// Formato UMD minimo (sin dependencias, sin build tools): funciona
// como <script> global en el navegador y como modulo CommonJS en
// Node (para las pruebas de paridad).
(function (global, factory) {
  if (typeof module !== "undefined" && module.exports) {
    module.exports = factory(require("./catalogo_datos.js"));
  } else {
    global.MOTOR = factory(global.CATALOGO_DATOS);
  }
})(typeof window !== "undefined" ? window : globalThis, function (CATALOGO_DATOS) {
  "use strict";

  var CORTES = CATALOGO_DATOS.cortes;
  var COMPATIBILIDADES = CATALOGO_DATOS.compatibilidades;
  var MECANISMOS = CATALOGO_DATOS.mecanismos;
  var EXCEPCIONES = CATALOGO_DATOS.excepciones;
  var ROSTROS = CATALOGO_DATOS.rostros;

  var VOLUMEN = { NINGUNO: 0, BAJO: 1, MEDIO: 2, ALTO: 3 };
  var EJE = { MUY_ANGULAR: 1, ANGULAR: 2, NEUTRO: 3, SUAVE: 4, MUY_SUAVE: 5 };
  var ZONA_A_CAMPO = { arriba: "volumenArriba", lateral: "volumenLateral", posterior: "volumenPosterior" };

  var MECANISMOS_POR_ID = {};
  MECANISMOS.forEach(function (m) { MECANISMOS_POR_ID[m.id] = m; });

  var COMPAT_POR_CLAVE = {};
  COMPATIBILIDADES.forEach(function (c) { COMPAT_POR_CLAVE[c.corteId + "|" + c.textura] = c; });

  var EXCEPCION_POR_CLAVE = {};
  EXCEPCIONES.forEach(function (e) { EXCEPCION_POR_CLAVE[e.corteId + "|" + e.textura] = e; });

  // -- FASE 1a: puntaje de rostro (0/1/2), condicional por modo --------

  function nivelParaObjetivoZona(objetivo, volumen) {
    if (objetivo === "sin_objetivo") return null;
    if (objetivo === "sumar") {
      if (volumen === VOLUMEN.ALTO) return 2;
      if (volumen === VOLUMEN.MEDIO) return 1;
      return 0; // BAJO o NINGUNO
    }
    if (objetivo === "reducir") {
      if (volumen === VOLUMEN.NINGUNO || volumen === VOLUMEN.BAJO) return 2;
      if (volumen === VOLUMEN.MEDIO) return 1;
      return 0; // ALTO
    }
    throw new Error("Objetivo de zona desconocido: " + objetivo);
  }

  function rostroPorZona(corte, definicion) {
    var niveles = [];
    Object.keys(definicion.objetivosZona).forEach(function (zona) {
      var campo = ZONA_A_CAMPO[zona];
      var nivel = nivelParaObjetivoZona(definicion.objetivosZona[zona], corte[campo]);
      if (nivel !== null) niveles.push(nivel);
    });
    if (niveles.length === 0) return 2;
    return Math.min.apply(null, niveles);
  }

  function rostroPorEje(corte, definicion) {
    if (corte.eje >= definicion.objetivoEjeMinimo) return 2;
    if (corte.eje === EJE.NEUTRO) return 1;
    return 0;
  }

  function rostroPorEquilibrio(corte) {
    var zonasEnAlto = [corte.volumenArriba, corte.volumenLateral, corte.volumenPosterior]
      .filter(function (v) { return v === VOLUMEN.ALTO; }).length;
    return zonasEnAlto === 3 ? 0 : 2;
  }

  function evaluarRostro(corte, formaRostro) {
    var definicion = ROSTROS[formaRostro];
    if (!definicion) throw new Error("Forma de rostro desconocida: " + formaRostro);
    if (definicion.modo === "por_zona") return rostroPorZona(corte, definicion);
    if (definicion.modo === "por_eje") return rostroPorEje(corte, definicion);
    if (definicion.modo === "por_equilibrio") return rostroPorEquilibrio(corte);
    throw new Error("Modo de evaluacion desconocido: " + definicion.modo);
  }

  // -- FASE 1b: puntaje de textura (lookup, nunca distancia) -----------

  function evaluarTextura(corte, textura) {
    return COMPAT_POR_CLAVE[corte.id + "|" + textura].valor;
  }

  // -- FASE 1c: deficit unilateral de densidad/grosor -------------------

  function nivelPorDeficit(referencia, real) {
    if (real >= referencia) return 2;
    var deficit = referencia - real;
    if (deficit === 1) return 1;
    return 0; // deficit === 2
  }

  function evaluarDensidad(corte, densidadCliente) {
    return nivelPorDeficit(corte.densidadReferencia, densidadCliente);
  }

  function evaluarGrosor(corte, grosorCliente) {
    return nivelPorDeficit(corte.grosorReferencia, grosorCliente);
  }

  // -- FASE 3: etiquetado SOLIDO / ACEPTABLE / DEBIL ---------------------

  function etiquetar(rostro, densidad, grosor) {
    var ambosEnCero = densidad === 0 && grosor === 0;
    var algunEjeEnCero = densidad === 0 || grosor === 0;
    if (rostro === 2 && !ambosEnCero) return "solido";
    if (rostro === 1 && algunEjeEnCero) return "debil";
    return "aceptable";
  }

  // -- Gates + evaluacion de candidatos ----------------------------------

  function evaluarCandidatos(diagnostico) {
    var candidatos = [];
    CORTES.forEach(function (corte) {
      var rostro = evaluarRostro(corte, diagnostico.formaRostro);
      if (rostro === 0) return; // gate: contradice el objetivo de visagismo
      var textura = evaluarTextura(corte, diagnostico.textura);
      if (textura === 0) return; // gate: imposible de ejecutar con esta textura
      var densidad = evaluarDensidad(corte, diagnostico.densidad);
      var grosor = evaluarGrosor(corte, diagnostico.grosor);
      var etiqueta = etiquetar(rostro, densidad, grosor);
      candidatos.push({ corte: corte, rostro: rostro, textura: textura, densidad: densidad, grosor: grosor, etiqueta: etiqueta });
    });
    return candidatos;
  }

  // -- FASE 4: comparador lexicografico completo -------------------------

  function claveOrden(candidato) {
    return [
      -candidato.rostro,
      -candidato.textura,
      -Math.min(candidato.densidad, candidato.grosor),
      -(candidato.densidad + candidato.grosor),
      candidato.corte.mantenimiento,
    ];
  }

  function compararCandidatos(a, b) {
    var claveA = claveOrden(a);
    var claveB = claveOrden(b);
    for (var i = 0; i < claveA.length; i++) {
      if (claveA[i] !== claveB[i]) return claveA[i] - claveB[i];
    }
    if (a.corte.nombre < b.corte.nombre) return -1;
    if (a.corte.nombre > b.corte.nombre) return 1;
    return 0;
  }

  function ordenar(candidatos) {
    return candidatos.slice().sort(compararCandidatos);
  }

  // -- FASE 6: explicacion y advertencias --------------------------------

  var NOMBRE_DENSIDAD = { 1: "baja", 2: "media", 3: "alta" };
  var NOMBRE_GROSOR = { 1: "fino", 2: "medio", 3: "grueso" };
  var NOMBRE_MANTENIMIENTO = { 1: "bajo", 2: "medio", 3: "alto" };

  var PLANTILLA_ADVERTENCIA = {
    0: "Advertencia fuerte: {articulo} {atributo} del cliente ({real}) esta muy por debajo " +
       "de la referencia de este corte ({referencia}); el resultado puede verse " +
       "notablemente comprometido.",
    1: "Advertencia leve: {articulo} {atributo} del cliente ({real}) esta un escalon " +
       "por debajo de la referencia de este corte ({referencia}); el corte es viable " +
       "con un ajuste tecnico menor.",
  };

  function formatear(plantilla, valores) {
    return plantilla.replace(/\{(\w+)\}/g, function (coincidencia, clave) {
      return Object.prototype.hasOwnProperty.call(valores, clave) ? valores[clave] : coincidencia;
    });
  }

  function calcularAdvertencias(corte, diagnostico, densidadScore, grosorScore) {
    var advertencias = [];
    if (densidadScore === 0 || densidadScore === 1) {
      advertencias.push(formatear(PLANTILLA_ADVERTENCIA[densidadScore], {
        articulo: "la",
        atributo: "densidad",
        real: NOMBRE_DENSIDAD[diagnostico.densidad],
        referencia: NOMBRE_DENSIDAD[corte.densidadReferencia],
      }));
    }
    if (grosorScore === 0 || grosorScore === 1) {
      advertencias.push(formatear(PLANTILLA_ADVERTENCIA[grosorScore], {
        articulo: "el",
        atributo: "grosor",
        real: NOMBRE_GROSOR[diagnostico.grosor],
        referencia: NOMBRE_GROSOR[corte.grosorReferencia],
      }));
    }
    return advertencias;
  }

  function generarExplicacion(candidato, diagnostico) {
    var corte = candidato.corte;
    var mecanismo = MECANISMOS_POR_ID[corte.mecanismoId];
    var excepcion = EXCEPCION_POR_CLAVE[corte.id + "|" + diagnostico.textura];
    return {
      nombre: corte.nombre,
      familia: corte.familia,
      etiqueta: candidato.etiqueta,
      efectos_visagismo: mecanismo.efectos.slice(),
      mecanismo: mecanismo.principioTecnico,
      nota_tecnica_textura: excepcion ? excepcion.texto : null,
      advertencias: calcularAdvertencias(corte, diagnostico, candidato.densidad, candidato.grosor),
      mantenimiento: NOMBRE_MANTENIMIENTO[corte.mantenimiento],
    };
  }

  // -- FASE 5: seleccion final (maximo 3, nunca fabricado) ----------------

  var MENSAJE_CATALOGO_VACIO = (
    "Ningun corte de la base de conocimiento actual cumple simultaneamente con " +
    "el objetivo de visagismo de este rostro y con la viabilidad de esta " +
    "textura/densidad/grosor. Se recomienda que el barbero aplique su criterio " +
    "profesional directamente; este caso queda registrado como un hueco de " +
    "cobertura del catalogo."
  );

  function recomendar(diagnostico) {
    var candidatos = evaluarCandidatos(diagnostico);
    if (candidatos.length === 0) {
      return { recomendaciones: [], mensaje_vacio: MENSAJE_CATALOGO_VACIO };
    }
    var ordenados = ordenar(candidatos);
    var top = ordenados.slice(0, 3);
    var recomendaciones = top.map(function (c) { return generarExplicacion(c, diagnostico); });
    return { recomendaciones: recomendaciones, mensaje_vacio: null };
  }

  return {
    evaluarRostro: evaluarRostro,
    evaluarTextura: evaluarTextura,
    evaluarDensidad: evaluarDensidad,
    evaluarGrosor: evaluarGrosor,
    etiquetar: etiquetar,
    evaluarCandidatos: evaluarCandidatos,
    ordenar: ordenar,
    generarExplicacion: generarExplicacion,
    recomendar: recomendar,
    MENSAJE_CATALOGO_VACIO: MENSAJE_CATALOGO_VACIO,
  };
});

File: src/motor/motor.py
"""Motor de recomendacion (grupo B: reglas).

Implementa, sin ninguna desviacion, la especificacion formal validada
con los 9 casos de prueba. No contiene datos del catalogo (eso vive en
`src.catalogo`) ni datos del cliente mas alla de lo que llega en
`DiagnosticoCliente`.

Determinista: la misma entrada produce siempre la misma salida.
Trazable: cada decision (gate, puntaje, etiqueta, orden) es una
funcion pura, independiente, que puede probarse por separado.
"""

from dataclasses import dataclass, field
from typing import Dict, List, Optional, Tuple

from ..catalogo.enums import (
    Densidad,
    Eje,
    Etiqueta,
    FormaRostro,
    Grosor,
    ModoEvaluacion,
    ObjetivoZona,
    Textura,
    Volumen,
    Zona,
)
from ..catalogo.compatibilidad import COMPATIBILIDAD_POR_CORTE_Y_TEXTURA
from ..catalogo.cortes import CORTES, Corte
from ..catalogo.excepciones import EXCEPCIONES_POR_CORTE_Y_TEXTURA
from ..catalogo.mecanismos import MECANISMOS_POR_ID
from ..catalogo.rostros import ROSTROS, DefinicionRostro
from ..diagnostico import DiagnosticoCliente


# ---------------------------------------------------------------------------
# FASE 1a -- Puntaje de rostro (0/1/2), condicional por modo de evaluacion.
# ---------------------------------------------------------------------------

def _volumen_de_zona(corte: Corte, zona: Zona) -> Volumen:
    return {
        Zona.ARRIBA: corte.volumen_arriba,
        Zona.LATERAL: corte.volumen_lateral,
        Zona.POSTERIOR: corte.volumen_posterior,
    }[zona]


def _nivel_para_objetivo_zona(objetivo: ObjetivoZona, volumen: Volumen) -> Optional[int]:
    """Tabla congelada de la especificacion (seccion 4.1)."""
    if objetivo == ObjetivoZona.SIN_OBJETIVO:
        return None
    if objetivo == ObjetivoZona.SUMAR:
        if volumen == Volumen.ALTO:
            return 2
        if volumen == Volumen.MEDIO:
            return 1
        return 0  # BAJO o NINGUNO
    if objetivo == ObjetivoZona.REDUCIR:
        if volumen in (Volumen.NINGUNO, Volumen.BAJO):
            return 2
        if volumen == Volumen.MEDIO:
            return 1
        return 0  # ALTO
    raise ValueError(f"Objetivo de zona desconocido: {objetivo}")


def _rostro_por_zona(corte: Corte, definicion: DefinicionRostro) -> int:
    niveles = [
        n for n in (
            _nivel_para_objetivo_zona(objetivo, _volumen_de_zona(corte, zona))
            for zona, objetivo in definicion.objetivos_zona.items()
        )
        if n is not None
    ]
    # El puntaje es el MINIMO entre las zonas con objetivo: un corte no
    # puede tapar con una zona excelente el hecho de contradecir el
    # objetivo en otra.
    return min(niveles) if niveles else 2


def _rostro_por_eje(corte: Corte, definicion: DefinicionRostro) -> int:
    umbral = definicion.objetivo_eje_minimo
    if corte.eje >= umbral:
        return 2
    if corte.eje == Eje.NEUTRO:
        return 1
    return 0


def _rostro_por_equilibrio(corte: Corte) -> int:
    zonas_en_alto = sum(
        1 for v in (corte.volumen_arriba, corte.volumen_lateral, corte.volumen_posterior)
        if v == Volumen.ALTO
    )
    # Unica condicion de ruptura del equilibrio: las 3 zonas en "alto"
    # a la vez, sin ningun punto de control. Con el catalogo actual
    # esto nunca ocurre (verificado en la validacion de 9 casos).
    return 0 if zonas_en_alto == 3 else 2


def evaluar_rostro(corte: Corte, forma: FormaRostro) -> int:
    definicion = ROSTROS[forma]
    if definicion.modo == ModoEvaluacion.POR_ZONA:
        return _rostro_por_zona(corte, definicion)
    if definicion.modo == ModoEvaluacion.POR_EJE:
        return _rostro_por_eje(corte, definicion)
    if definicion.modo == ModoEvaluacion.POR_EQUILIBRIO:
        return _rostro_por_equilibrio(corte)
    raise ValueError(f"Modo de evaluacion desconocido: {definicion.modo}")


# ---------------------------------------------------------------------------
# FASE 1b -- Puntaje de textura (0/1/2): lookup directo, nunca distancia.
# ---------------------------------------------------------------------------

def evaluar_textura(corte: Corte, textura: Textura) -> int:
    return COMPATIBILIDAD_POR_CORTE_Y_TEXTURA[(corte.id, textura)].valor


# ---------------------------------------------------------------------------
# FASE 1c -- Puntaje de densidad y grosor: deficit unilateral respecto a la
# referencia. "Mas de lo requerido" nunca penaliza.
# ---------------------------------------------------------------------------

def _nivel_por_deficit(referencia: int, real: int) -> int:
    if real >= referencia:
        return 2
    deficit = referencia - real
    if deficit == 1:
        return 1
    return 0  # deficit == 2, unico otro caso posible en una escala de 3 niveles


def evaluar_densidad(corte: Corte, densidad_cliente: Densidad) -> int:
    return _nivel_por_deficit(int(corte.densidad_referencia), int(densidad_cliente))


def evaluar_grosor(corte: Corte, grosor_cliente: Grosor) -> int:
    return _nivel_por_deficit(int(corte.grosor_referencia), int(grosor_cliente))


# ---------------------------------------------------------------------------
# FASE 3 -- Etiquetado SOLIDO / ACEPTABLE / DEBIL (tabla congelada y
# verificada exhaustiva sobre las 18 combinaciones posibles).
# ---------------------------------------------------------------------------

def etiquetar(rostro: int, densidad: int, grosor: int) -> Etiqueta:
    ambos_en_cero = densidad == 0 and grosor == 0
    algun_eje_en_cero = densidad == 0 or grosor == 0

    if rostro == 2 and not ambos_en_cero:
        return Etiqueta.SOLIDO
    if rostro == 1 and algun_eje_en_cero:
        return Etiqueta.DEBIL
    return Etiqueta.ACEPTABLE


# ---------------------------------------------------------------------------
# Candidato: resultado intermedio de evaluar un corte contra un diagnostico.
# ---------------------------------------------------------------------------

@dataclass(frozen=True)
class Candidato:
    corte: Corte
    rostro: int
    textura: int
    densidad: int
    grosor: int
    etiqueta: Etiqueta


def evaluar_candidatos(diagnostico: DiagnosticoCliente) -> List[Candidato]:
    """FASE 1 + FASE 2 (gates). Recorre los 15 cortes y descarta los que
    no pasan los UNICOS dos gates (rostro=0, textura=0)."""
    candidatos: List[Candidato] = []
    for corte in CORTES:
        rostro = evaluar_rostro(corte, diagnostico.forma_rostro)
        if rostro == 0:
            continue  # gate: contradice el objetivo de visagismo
        textura = evaluar_textura(corte, diagnostico.textura)
        if textura == 0:
            continue  # gate: imposible de ejecutar con esta textura
        densidad = evaluar_densidad(corte, diagnostico.densidad)
        grosor = evaluar_grosor(corte, diagnostico.grosor)
        etiqueta = etiquetar(rostro, densidad, grosor)
        candidatos.append(Candidato(corte, rostro, textura, densidad, grosor, etiqueta))
    return candidatos


# ---------------------------------------------------------------------------
# FASE 4 -- Comparador lexicografico completo.
# rostro > textura > min(densidad,grosor) > suma(densidad+grosor)
#        > mantenimiento (asc) > nombre (alfabetico, ultimo recurso).
# ---------------------------------------------------------------------------

def _clave_orden(c: Candidato) -> Tuple:
    return (
        -c.rostro,
        -c.textura,
        -min(c.densidad, c.grosor),
        -(c.densidad + c.grosor),
        int(c.corte.mantenimiento),
        c.corte.nombre,
    )


def ordenar(candidatos: List[Candidato]) -> List[Candidato]:
    return sorted(candidatos, key=_clave_orden)


# ---------------------------------------------------------------------------
# FASE 6 -- Explicacion y advertencias.
# ---------------------------------------------------------------------------

_PLANTILLA_ADVERTENCIA = {
    0: "Advertencia fuerte: {articulo} {atributo} del cliente ({real}) esta muy por "
       "debajo de la referencia de este corte ({referencia}); el resultado puede "
       "verse notablemente comprometido.",
    1: "Advertencia leve: {articulo} {atributo} del cliente ({real}) esta un escalon "
       "por debajo de la referencia de este corte ({referencia}); el corte es viable "
       "con un ajuste tecnico menor.",
}


def _advertencias(corte: Corte, diagnostico: DiagnosticoCliente,
                   densidad_score: int, grosor_score: int) -> List[str]:
    advertencias = []
    if densidad_score in (0, 1):
        advertencias.append(_PLANTILLA_ADVERTENCIA[densidad_score].format(
            articulo="la",
            atributo="densidad",
            real=diagnostico.densidad.name.lower(),
            referencia=corte.densidad_referencia.name.lower(),
        ))
    if grosor_score in (0, 1):
        advertencias.append(_PLANTILLA_ADVERTENCIA[grosor_score].format(
            articulo="el",
            atributo="grosor",
            real=diagnostico.grosor.name.lower(),
            referencia=corte.grosor_referencia.name.lower(),
        ))
    return advertencias


@dataclass(frozen=True)
class Recomendacion:
    corte_id: int
    nombre: str
    familia: str
    etiqueta: Etiqueta
    efectos_atendidos: List[str]
    mecanismo: str
    nota_tecnica_textura: Optional[str]
    advertencias: List[str]
    mantenimiento: str


def generar_explicacion(candidato: Candidato, diagnostico: DiagnosticoCliente) -> Recomendacion:
    corte = candidato.corte
    mecanismo = MECANISMOS_POR_ID[corte.mecanismo_id]
    excepcion = EXCEPCIONES_POR_CORTE_Y_TEXTURA.get((corte.id, diagnostico.textura))
    return Recomendacion(
        corte_id=corte.id,
        nombre=corte.nombre,
        familia=corte.familia,
        etiqueta=candidato.etiqueta,
        efectos_atendidos=[e.value for e in mecanismo.efectos],
        mecanismo=mecanismo.principio_tecnico,
        nota_tecnica_textura=excepcion.texto if excepcion else None,
        advertencias=_advertencias(corte, diagnostico, candidato.densidad, candidato.grosor),
        mantenimiento=corte.mantenimiento.name.lower(),
    )


# ---------------------------------------------------------------------------
# FASE 5 -- Seleccion final (maximo 3, nunca fabricado) + mensaje de vacio.
# ---------------------------------------------------------------------------

MENSAJE_CATALOGO_VACIO = (
    "Ningun corte de la base de conocimiento actual cumple simultaneamente con "
    "el objetivo de visagismo de este rostro y con la viabilidad de esta "
    "textura/densidad/grosor. Se recomienda que el barbero aplique su criterio "
    "profesional directamente; este caso queda registrado como un hueco de "
    "cobertura del catalogo."
)


@dataclass(frozen=True)
class ResultadoRecomendacion:
    recomendaciones: List[Recomendacion] = field(default_factory=list)
    mensaje_vacio: Optional[str] = None


def recomendar(diagnostico: DiagnosticoCliente) -> ResultadoRecomendacion:
    candidatos = evaluar_candidatos(diagnostico)
    if not candidatos:
        return ResultadoRecomendacion(recomendaciones=[], mensaje_vacio=MENSAJE_CATALOGO_VACIO)

    ordenados = ordenar(candidatos)
    top = ordenados[:3]
    recomendaciones = [generar_explicacion(c, diagnostico) for c in top]
    return ResultadoRecomendacion(recomendaciones=recomendaciones, mensaje_vacio=None)
